core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServices.java

```diff
@@ -108,9 +111,52 @@
      * Returns the {@link PathMapping} instance mapped to {@code method}.
      */
     private static PathMapping pathMapping(String pathPrefix, Method method) {
-        Path mapping = method.getAnnotation(Path.class);
-        String mappedTo = ArmeriaHttpUtil.concatPaths(pathPrefix, mapping.value());
-        return PathMapping.of(mappedTo);
+        pathPrefix = ensureAbsolutePath(pathPrefix, "pathPrefix");
+        if (!pathPrefix.endsWith("/")) {
+            pathPrefix += '/';
+        }
+
+        final String pattern = method.getAnnotation(Path.class).value();
+        if (pattern == null) {
+            throw new NullPointerException("The value of @Path must be non-null.");
+        }
+
+        final PathMapping mapping = PathMapping.of(pattern);
+        if ("/".equals(pathPrefix)) {
+            // pathPrefix is not specified or "/".
+            return mapping;
+        }
+
+        if (pattern.startsWith("exact:")) {
+            // 6 = "exact:".length()
```

Comment:
Can inline `exact.length` into the code instead of writing comments (JIT inlines the constant for equal performance)